src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -937,35 +948,33 @@ static int dhcp_server_relay_message(sd_dhcp_server *server, DHCPMessage *messag
         return -EBADMSG;
 }
 
-static int prepare_new_lease(
-                DHCPLease **ret_lease,
-                be32_t address,
-                const DHCPClientId *client_id,
-                uint8_t htype,
-                uint8_t hlen,
-                const uint8_t *chaddr,
-                be32_t gateway,
-                usec_t expiration) {
+static int prepare_new_lease(DHCPLease **ret_lease, be32_t address, DHCPRequest *req, usec_t expiration) {
 
         _cleanup_(dhcp_lease_freep) DHCPLease *lease = NULL;
 
         lease = new(DHCPLease, 1);
         if (!lease)
                 return -ENOMEM;
 
-        *lease = (DHCPLease) {
+        *lease = (DHCPLease){
                 .address = address,
-                .client_id.length = client_id->length,
-                .htype = htype,
-                .hlen = hlen,
-                .gateway = gateway,
+                .client_id.length = req->client_id.length,
+                .htype = req->message->htype,
+                .hlen = req->message->hlen,
+                .gateway = req->message->giaddr,
                 .expiration = expiration,
         };
-        lease->client_id.data = memdup(client_id->data, client_id->length);
+        lease->client_id.data = memdup(req->client_id.data, req->client_id.length);
         if (!lease->client_id.data)
                 return -ENOMEM;
 
-        memcpy(lease->chaddr, chaddr, hlen);
+        if (req->hostname) {
+                lease->hostname = strdup(req->hostname);
+                if (!lease->hostname)
+                        return -ENOMEM;
+        }
+
+        memcpy(lease->chaddr, req->message->chaddr, req->message->hlen);
```

Comment:
Not a strong opinion, but I'd like to copy chaddr first, as the hostname is optional.

Reply:
Done.